### PR TITLE
Add new update groups handler and associated logic.

### PR DESCRIPTION
**Pull request by mike-kelly-methods**
### What

As per title. Add new API endpoint to update a group's name (description) and precedence.

Group precedence is now `> 10` and `< 100`.

Add new API, Model and Feature tests to support functional logic.

### How to review

Tests and Build runs without errors. Check code and sanitise.

### Who can review

Anyone but me.

### CODE REVIEW

File: api/groups_test.go [hunk 1]
```diff
@@ -767,16 +789,46 @@ func TestCreateNewGroup(t *testing.T) {
 					So(errorResponse, ShouldBeNil)
 				},
 			},
-			// 400 response - group already exists
```
Comment (AWitcherONS) [inline api/groups_test.go:770]: I don't understand why this test was removed 
surely that a test for whether a group already exists is valid 
and that response is more meaningful to a user...

**Comment by AWitcherONS** [comment]
Shouldn't there be a swagger for PUT groups?
there were a few gofmt but nothing to fail 
test pass coverage of commit 
test-component passed

File: api/api.go [hunk 2]
```diff
@@ -91,6 +91,7 @@ func Setup(ctx context.Context, r *mux.Router, cognitoClient cognito.Client, use
 	r.HandleFunc("/v1/groups", contextAndErrors(api.ListGroupsHandler)).Methods(http.MethodGet)
 	r.HandleFunc("/v1/groups", contextAndErrors(api.CreateGroupHandler)).Methods(http.MethodPost)
 	r.HandleFunc("/v1/groups/{id}", contextAndErrors(api.GetGroupHandler)).Methods(http.MethodGet)
+	r.HandleFunc("/v1/groups/{id}", contextAndErrors(api.UpdateGroupHandler)).Methods(http.MethodPut)
```
Comment (AWitcherONS) [inline api/api.go:94]: api_test.go should have unit test to cover new endpoint....
So(hasRoute(api.Router, "/v1/groups/{id}", http.MethodPut), ShouldBeTrue)

**Review by AWitcherONS — DISMISSED** [review]

File: api/api.go (hunk 2 above)
Comment (mike-kelly-methods) [reply api/api.go:94]: Added.

File: models/groups.go [hunk 3]
```diff
@@ -148,39 +149,62 @@ func (g *Group) BuildSuccessfulJsonResponse(ctx context.Context) ([]byte, error)
...
 	var validationErrs []error
 
 	if g.Description == nil {
 		validationErrs = append(validationErrs, NewValidationError(ctx, InvalidGroupName, MissingGroupName))
 	} else if m, _ := regexp.MatchString("(?i)^role_.*", *g.Description); m {
 		validationErrs = append(validationErrs, NewValidationError(ctx, InvalidGroupName, IncorrectPatternInGroupName))
+	} else {
+		//ensure groups name in description doesn't already exist - creation only
+		if g.GroupsList != nil {
+			g.CleanGroupDescription()
```
Comment (mwnicks) [inline models/groups.go]: Are we still cleaning the Group description?
I thought it was meant to be human readable now, as we have the UUID for urls etc.

File: api/groups.go [hunk 4]
```diff
@@ -20,21 +20,32 @@ func (api *API) CreateGroupHandler(ctx context.Context, w http.ResponseWriter, r
...
+	}
+
+	validationErrs := createGroup.ValidateCreateUpdateGroupRequest(ctx)
 	if len(validationErrs) != 0 {
 		return nil, models.NewErrorResponse(http.StatusBadRequest, nil, validationErrs...)
 	}
 
-	// build create group input
+	// remove special chars from description
 	createGroup.CleanGroupDescription()
```
Comment (mwnicks) [inline api/groups.go]: Isn't this already run in the validation step?
(although see comment in the validation step about if this is needed)

File: swagger.yaml [hunk 5]
```diff
@@ -467,6 +467,46 @@ paths:
...
+      consumes:
+        - application/json
+      produces:
+        - "application/json"
+      parameters:
+        - in: body
+          name: "Group"
+          description: "The updated details of the group"
+          schema:
+            required: ["name", "precedence"]
```
Comment (mwnicks) [inline swagger.yaml:489]: Should this not be 'description'?

File: swagger.yaml [hunk 6]
```diff
@@ -467,6 +467,46 @@ paths:
...
+            type: object
+            properties:
+              name:
```
Comment (mwnicks) [inline swagger.yaml:492]: Should this not be 'description'?

File: swagger.yaml [hunk 7]
```diff
@@ -467,6 +467,46 @@ paths:
...
+    put:
+      tags:
+        - Groups
+      summary: "Update an existing group"
+      description: "Update an existing group in Cognito user pool"
+      consumes:
+        - application/json
+      produces:
+        - "application/json"
+      parameters:
```
Comment (mwnicks) [inline swagger.yaml:479]: Seems to be missing the `- in: path` parameter for the group name

File: api/groups.go (hunk 4 above)
Comment (mike-kelly-methods) [reply api/groups.go]: The `CleanGroupDescription()` function is applied to the incoming data after the validation step - as calling it before the validation, and without checking if the description is populated, would result in a null pointer exception.

File: models/groups.go (hunk 3 above)
Comment (mike-kelly-methods) [reply models/groups.go]: I wanted to double check this - thats fine, I can remove the regex replace on special charset then.

File: swagger.yaml (hunk 5 above)
Comment (mike-kelly-methods) [reply swagger.yaml:489]: Changed this to `name` as per ticket https://trello.com/c/i7MqRBIg/670-670-add-human-readable-group-name:

`API Req -> Cognito - > API resp`
`name -> Description -> name`

File: swagger.yaml (hunk 6 above)
Comment (mike-kelly-methods) [reply swagger.yaml:492]: As per above.

**Review by AWitcherONS — COMMENTED** [review]
make tests pass 
make test-component pass
gofmt show format warnings but nothing to stop release
Postman suite for dp-identity-api run and all pass
no new Print statements

**Review by mwnicks — APPROVED** [review]
LGTM 👍

**Pull request merged**